src/Headers/Ic.h

```diff
@@ -104,7 +105,7 @@ class Ic
   virtual void Generate(void) {
     ExceptionHandler::getIstance().raise("Ic::Generate function not implemented");
   };
-  virtual FLOAT GetDensity(const FLOAT *) const {
+  virtual FLOAT GetDensity(const FLOAT *, const Typemask typemask=Typemask(true)) const {
```

Comment:
> Are we sure that using a Typemask is the right way to go here: It's definitely the most flexible so I won't argue on that ground, but it is a bit of a pain to use when we want just a single particle type. The other option would be to make GetDensity take a ptype. We'd then just call it multiple times if we wanted a 'total density' over some set of particles.

Yes, this is the main thing I realised after committing.  It's probably best to give it just the particle type (btw, on a slight tangent, but we have both ptype and parttype.  Do we really need both?  Any way to get rid of one because it seems overkill to have two.)  I can make this change now also.

> By the way, these functions should not be virtual.

How can be they not be virtual??  I mean, we are declaring them as pure virtual to force an implementation when they are needed so I'm not sure how else this can be done.  Unless there's some C++ wizardry I'm unaware of!  If they can be inlined then great, since that should speed it up a little.